core/go/internal/publictxmgr/in_flight_transaction_stage_controller.go

```diff
@@ -195,256 +204,102 @@ func (it *inFlightTransactionStageController) ProduceLatestInFlightStageContext(
+			it.stateManager.UpdateTransaction(update)
+			madeUpdate = true
+		}
+	}
+
+	if madeUpdate {
+		// If we have made an update we don't wait to collect the output of whatever stages might be already running before starting
+		// the process of submitting the transaction with its new values. We track any stages that are still running to make sure
+		// important parts are persisted but this doesn't stop stages with the new values from starting.
+		it.stateManager.NewVersion(ctx)
+	}
+
 	// update the transaction orchestrator context
 	it.stateManager.SetOrchestratorContext(ctx, tIn)
-	if it.stateManager.GetRunningStageContext(ctx) != nil {
-		rsc := it.stateManager.GetRunningStageContext(ctx)
-		log.L(ctx).Debugf("ProduceLatestInFlightStageContext for tx %s, on stage: %s , current stage context lived: %s , stage lived: %s, last stage error: %+v", it.stateManager.GetSignerNonce(), it.stateManager.GetStage(ctx), time.Since(rsc.StageStartTime), time.Since(it.stateManager.GetStageStartTime(ctx)), it.stateManager.GetStageTriggerError(ctx))
-		// once we have a running context, all the metadata should already be loaded
-		if it.stateManager.GetStageTriggerError(ctx) != nil {
-			log.L(ctx).Errorf("Failed to trigger sate due to %+v, cleaning up the context and retry", it.stateManager.GetStageTriggerError(ctx))
-			it.stateManager.ClearRunningStageContext(ctx)
+
+	// most outputs of a previous version can be ignored but some (submission persistence) need to be processed
+	// even if we've moved onto trying to submit a new version
+	it.processPreviousVersionsStateOutputs(ctx)
+
+	tOut.Error = it.processCurrentVersionStateOutputs(ctx)
+
+	if it.stateManager.GetGasPriceObject() != nil {
+		if it.stateManager.IsReadyToExit() {
+			// already has confirmed transaction so the cost to submit this transaction is zero
+			tOut.Cost = big.NewInt(0)
 		} else {
-			// there is a running stage waiting for inputs
-			// first of checking the inputs to see whether we have new items to process
-			it.stateManager.ProcessStageOutputs(ctx, func(stageOutputs []*StageOutput) (unprocessedStageOutputs []*StageOutput) {
+			gpo := it.stateManager.GetGasPriceObject()
+			c, err := calculateGasRequiredForTransaction(ctx, gpo, it.stateManager.GetGasLimit())
+			if err == nil {
+				tOut.Cost = c
+			}
+		}
+	}
+
+	// Only the current version is progressed by starting a new stage
+	if it.stateManager.GetCurrentVersion(ctx).GetRunningStageContext(ctx) == nil {
+		// no running context in flight
+		// The action for each stage can be started asynchronously; however, any transation values from the in memory transaction must
+		// be read within this goroutine so that we know that they haven't been changed by an update part way through.
+		it.startNewStage(ctx, tOut.Cost)
+	}
+	tOut.TransactionSubmitted = it.stateManager.GetTransactionHash() != nil
+
+	return tOut
+}
+
+func (it *inFlightTransactionStageController) processCurrentVersionStateOutputs(ctx context.Context) (err error) {
```

Comment:
Bit hard for me to process what's changed int his function, but I'm very happy to see a lot of removed nested code.

This function has felt like to me one of the hardest to understand / page into bits of the code in Paladin, as understanding it relies on a strong understanding of really complex constructs (like "stage outputs", or even "stages") that aren't really called out in the structure of the code.

So a move towards a codebase where these things are more self-describing, and contained together, feels like it will really help new maintainers (and people like me that tent to page in/out of this code, rather than being able to hold the naming/structure in my head the whole time).